test/bob/execution/internals_test.clj

```diff
@@ -49,62 +54,130 @@
 
   (testing "unsuccessful pull"
     (with-redefs-fn {#'has-image   (constantly (f/fail "Failed"))
-                     #'docker/pull (fn [_ _] (throw (Exception. "Failed")))}
-      #(let [result (pull "img")]
+                     #'docker/invoke (fn [_ _] {:message "failed"})}
+      #(let [result (pull-image "clojure:foo")]
          (is (and (f/failed? result)
-                  (= "Cannot pull img"
-                     (f/message result))))))))
+                  (= "Could not pull image: clojure:foo"
+                     (f/message result)))))))
+
+  (testing "missing tag in image name"
+    (with-redefs-fn {#'docker/invoke (constantly {:message "failed"})}
+      #(let [result (pull-image "clojure")]
+         (is (f/failed? result)))))
+
+  (testing "image already locally available"
+    (with-redefs-fn {#'has-image   (constantly true)}
+      #(is (= "img" (pull-image "img"))))))
+
+(deftest container-build
+  (testing "successful build"
+    (with-redefs-fn {#'docker/invoke (constantly {:Id "83996a1d", :Warnings []})}
+      #(is (= "83996a1d" (create-container "foo:bar"
+                                           {:needs_resource "source"
+                                            :cmd "ls"}
+                                           {:FOO "bar"})))))
+
+  (testing "successful build single param"
+    (with-redefs-fn {#'docker/invoke (constantly {:Id "83996a1d", :Warnings []})}
+      #(is (= "83996a1d" (create-container "foo:bar")))))
+
+  (testing "successful build double param"
+    (with-redefs-fn {#'docker/invoke (constantly {:Id "83996a1d", :Warnings []})}
+      #(is (= "83996a1d" (create-container "foo:bar" {:cmd "ls"})))))
+
+  (testing "failing build"
+    (with-redefs-fn {#'docker/invoke (constantly {:message "failed"})}
+      #(is (f/failed? (create-container "foo:bar"
+                                        {:needs_resource "source"
+                                         :cmd "ls"}
+                                        {:FOO "bar"}))))))
 
 (deftest container-status
   (testing "successful status fetch"
-    (with-redefs-fn {#'docker/container-state (fn [_ id]
-                                                (tu/check-and-fail
-                                                 #(= "id" id))
-                                                {:Running  false
-                                                 :ExitCode 0})}
+    (with-redefs-fn {#'docker/invoke (constantly {:State {:Running  false
+                                                          :ExitCode 0}})}
       #(is (= {:running? false :exit-code 0} (status-of "id")))))
 
   (testing "unsuccessful status fetch"
-    (with-redefs-fn {#'docker/container-state (fn [_ _] (throw (Exception. "Failed")))}
+    (with-redefs-fn {#'docker/invoke (fn [_ id] {:message (format "No such container: %s" id)})}
       #(is (f/failed? (status-of "id"))))))
 
-(deftest container-runs
-  (testing "successful run"
+;(deftest logs-live-test
```

Comment:
We can remove this?

Reply:
I tried to test the streaming of logs but failed. Should we pursue that?

Reply:
we can remove this. its again side effects. maybe just leave a TODO here without the comment?

Reply:
I removed the whole comment. integration tests are covering it and since it is side effects anyways we are better off with integration testing it... I think...